amd_tressfx_vulkan/src/TressFXSimulationVulkan.cpp

```diff
@@ -0,0 +1,1180 @@
+                            AMD_ARRAY_SIZE(lengthWindCollisionSets),
+                            lengthWindCollisionSets, AMD_ARRAY_SIZE(descriptorOffset),
+                            descriptorOffset);
+    vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                      m_CSLengthConstriantsWindAndCollision);
+    vkCmdDispatch(commandBuffer, numOfGroupsForCS_VertexLevel, 1, 1);
+
+    VkBufferMemoryBarrier lengthWindCollisionBarriers[] = {
+        getBufferBarrier(m_pTressFXMesh->m_HairVertexPositionsBuffer,
+                         VK_ACCESS_SHADER_WRITE_BIT | VK_ACCESS_SHADER_READ_BIT,
+                         VK_ACCESS_MEMORY_WRITE_BIT | VK_ACCESS_SHADER_READ_BIT)};
+
+    vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                         VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0, 0, nullptr,
+                         AMD_ARRAY_SIZE(lengthWindCollisionBarriers),
+                         lengthWindCollisionBarriers, 0, nullptr);
+
+    // Update follow hair vertices
+    // One thread computes one vertex
+    if (m_bGuideFollowHairPrev)
+    {
+        VkDescriptorSet updateFollowHairSets[] = {m_configSet,
+                                                  m_pTressFXMesh->m_UpdateFollowHairSet};
+        vkCmdBindDescriptorSets(
+            commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+            m_CSUpdateFollowHairVerticesLayout, 0, AMD_ARRAY_SIZE(updateFollowHairSets),
+            updateFollowHairSets, AMD_ARRAY_SIZE(descriptorOffset), descriptorOffset);
+        vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                          m_CSUpdateFollowHairVertices);
+        vkCmdDispatch(commandBuffer, numOfGroupsForCS_VertexLevel, 1, 1);
+
+        VkBufferMemoryBarrier updateFollowHairBarriers[] = {
+            getBufferBarrier(m_pTressFXMesh->m_HairVertexPositionsBuffer,
+                             VK_ACCESS_SHADER_WRITE_BIT | VK_ACCESS_SHADER_READ_BIT,
+                             VK_ACCESS_SHADER_READ_BIT),
+        };
+
+        vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                             VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0, 0, nullptr,
+                             AMD_ARRAY_SIZE(updateFollowHairBarriers),
+                             updateFollowHairBarriers, 0, nullptr);
+    }
+
+    VkBufferMemoryBarrier makeTangentBarrier[] = {getBufferBarrier(
+        m_pTressFXMesh->m_HairVertexTangentsBuffer, VK_ACCESS_MEMORY_READ_BIT,
+        VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT)};
+
+    vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                         VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0, 0, nullptr,
+                         AMD_ARRAY_SIZE(makeTangentBarrier), makeTangentBarrier, 0,
+                         nullptr);
+
+    // Compute tangents for every vertex (guide + follow)
+    {
```

Comment:
Move this block into a separate function - makes it clearer what is happening there, and you can remove the comment.